Step3 legacy refactoring

mwkwon opened the pull request:
안녕하세요~ 리뷰어님~!!!
항상 꼼꼼하게 리뷰 해주셔서 감사합니다~^^
이번 리뷰도 피가되고 살이될 수 있도록 뼈때리는 신랄한 리뷰 부탁 드립니다.

질문 사항이 있는데 QnaServiceTest 를 실행하려는데 
```
org.mockito.exceptions.misusing.UnnecessaryStubbingException: 
Unnecessary stubbings detected.
Clean & maintainable test code requires zero unnecessary code.
Following stubbings are unnecessary (click to navigate to relevant line of code):
  1. -> at qna.service.QnaServiceTest.delete_답변_중_다른_사람이_쓴_글(QnaServiceTest.java:86)
Please remove unnecessary stubbings or use 'lenient' strictness. More info: javadoc for UnnecessaryStubbingException class.
```

아래와 같은 에러가 발생하여 관련 내용 구글에서 검색해보니
`@MockitoSettings(strictness = Strictness.LENIENT)`
어노테이션을 추가하라고 나오더라구요.
mock 객체 생성 시 불필요한 stubbings 한 것이 원인이라고하는데 어떤 내용인지... 제가 잘 이해를 못하겠어요..
해당 어노테이션을 안쓰게 좋다고 하는데 어떻게 처리해야할지 감이 안오네요..ㅠㅠ
관련 내용 확인 부탁 드릴께요~!

항상 감사합니다~ ^^

Commit:
refactor: 3단계 시작 전 소스 리펙토링

- createdAt, updateAt jpaAuditing 적용
- repository test 코드 작성
- 불필요한 method 제거
- cascade 적용

Commit:
docs: 기능 구현 목록 작성

기능 구현 목록 최초 작성

Commit:
feat: 로그인 사용자와 질문한 사람이 같은지 비교 기능

로그인 사용자와 질문한 사람이 다른 경우 CannotDeleteException 발생 처리
user class isSameUser method 추가

Commit:
feat: answers 일급 컬렉션 생성

질문에 답변이 존재하는지 여부 확인 기능 구현
질문에 대한 답변이 존재하는 경우 질문자와 답변자의 사용자가 같은지 유효성 검사 기능 구현

mwkwon committed:
docs: 기능 구현 목록 수정

mwkwon committed:
feat: 원시값 포장 객체 생성

title, contents, deleted, name, email 원시값 포장 객체 생성
validationUtil class 생성
관련 테스트 코드 작성

mwkwon committed:
feat: 삭제 히스토리 생성 기능

질문 삭제 히스토리 기능 구현
답변 삭제 히스토리 기능 구현
관련 테스트 코드 작성

kyucumber commented on src/test/java/qna/service/QnaServiceTest.java:
```diff
@@ -20,6 +22,7 @@
 import static org.mockito.Mockito.when;
 
 @ExtendWith(MockitoExtension.class)
+@MockitoSettings(strictness = Strictness.LENIENT)
```
```
@MockitoSettings(strictness = Strictness.LENIENT)
어노테이션을 추가하라고 나오더라구요.
mock 객체 생성 시 불필요한 stubbings 한 것이 원인이라고하는데 어떤 내용인지... 제가 잘 이해를 못하겠어요..
해당 어노테이션을 안쓰게 좋다고 하는데 어떻게 처리해야할지 감이 안오네요..ㅠㅠ
관련 내용 확인 부탁 드릴께요~!
```

`Mockito`와 테스트에 대한 개념이 아직 완전히 잡히지 않으신 것으로 보이네요.

기존 작성해주신 아래 테스트를 예시로 설명드리겠습니다.

```java
    @Test
    public void delete_성공() throws Exception {
        when(questionRepository.findByIdAndDeletedFalse(question.getId())).thenReturn(Optional.of(question));
        when(answerRepository.findByQuestionIdAndDeletedFalse(question.getId())).thenReturn(Arrays.asList(answer)); // 불필요한 부분

        assertThat(question.isDeleted()).isFalse();
        qnaService.deleteQuestion(UserTest.JAVAJIGI, question.getId());

        assertThat(question.isDeleted()).isTrue();
        verifyDeleteHistories();
    }
```

`qnaService.deleteQuestion(` 호출 시 아래 mocking 해주신 부분은 호출이 되지 않습니다.
mocking할 필요가 없는 메소드라는 의미가 되겠네요.

```java
answerRepository.findByQuestionIdAndDeletedFalse(question.getId()); 
```

따라서 불필요하게 스터빙을 했다고 예외가 발생하는 것 같습니다. 

이 부분은 명확하게 필요한 부분만 스터빙 하도록 변경해주셔야 합니다. `@MockitoSettings(strictness = Strictness.LENIENT)`와 같은 어노테이션을 붙여 우회해 테스트를 통과하도록 구현하는것은 좋지 않을 것 같네요.

mwkwon replied:
감사합니다~^^

kyucumber commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -2,30 +2,33 @@
     private Long id;
 
-    @ManyToOne(fetch = FetchType.LAZY)
+    @ManyToOne(fetch = FetchType.LAZY, cascade = CascadeType.PERSIST)
```
이 부분은 왜 추가하신걸까요?

`Answer` 작성 시 유저에 영속성 전이가 필요할까요?

kyucumber commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -2,30 +2,33 @@
     private User writer;
 
-    @ManyToOne(fetch = FetchType.LAZY)
+    @ManyToOne(fetch = FetchType.LAZY, cascade = CascadeType.PERSIST)
```
`Answer`보다 `Question`이 우선적으로 생성되는 경우가 많을 것 같은데 영속성 전이가 왜 필요한지 궁금하네요.

의도를 남겨주시면 감사하겠습니다.

mwkwon replied:
테스트 코드 작성 시 save() method를 사용하지 않으려고 추가하였습니다.
다시 보니 리뷰어님 말씀대로 불필요했던 것으로 생각됩니다.
감사합니다.

kyucumber commented on src/main/java/qna/domain/DeleteHistory.java:
```diff
@@ -15,32 +19,47 @@
+    private ContentId contentId = new ContentId();
 
-    @ManyToOne
+    @ManyToOne(fetch = FetchType.LAZY, cascade = CascadeType.PERSIST)
```
이 부분도 확인이 필요할 것 같네요.

kyucumber commented on src/main/java/qna/domain/Question.java:
```diff
@@ -38,67 +48,82 @@ public Question(String title, String contents) {
-        this.deleted = deleted;
+    private DeleteHistory createDeleteHistory(User loginUser) {
+        return DeleteHistory.create(QUESTION_CONTENT_TYPE, id, loginUser);
     }
```
`Question`이 `DeleteHistory`에 대한 생성 등의 책임을 가지고 있네요.

`Question`이 `DeleteHistory`와 결합하는 부분이 좋은지 조금 고민이 필요할 것 같습니다.

`Question`, `Answer` 등 늘어나는 모든 엔티티들의 삭제 기록을 남겨야 하면 모든 엔티티가 `DeleteHistory`를 참조하게 될 텐데 그렇게 되면 `DeleteHistory`의 변경에 모든 엔티티가 영향을 받게 되겠네요.

kyucumber commented on src/test/java/qna/domain/AnswerTest.java:
```diff
@@ -1,112 +1,22 @@
-        Answer expected = answers.save(answer);
-        entityManager.flush();
-        entityManager.clear();
+    void 답변_작성자가_로그인_사용와_같은지_비교() {
+        Answer answer = AnswerTest.A1;
 
-        Optional<Answer> actual = answers.findById(expected.getId());
-        assertThat(actual.isPresent()).isTrue();
-        assertThat(actual.get().getWriter().getId()).isEqualTo(user.getId());
-    }
+        User user = UserTest.JAVAJIGI;
+        User other = UserTest.SANJIGI;
```
테스트 클래스에서 다른 테스트의 static 변수를 사용하는 부분을 제거해주세요.

테스트는 외부 요인에 의해 영향을 받지 않도록 구성해주셔야 합니다.

kyucumber reviewed (changes requested):
안녕하세요

몇가지 코멘트 남겨두었으니 확인 부탁드립니다.

mwkwon committed:
refactor: 코드 리뷰 코멘트 사항 반영

테스트 코드의 다른 클래스 static 변수 제거
불필요한 casecade 제거
불필요한 스터빙 제거
DeleteHistory 생성 로직 변경

kyucumber commented on src/test/java/qna/domain/QuestionTest.java:
```diff
@@ -1,109 +1,48 @@
 
-@DataJpaTest
 public class QuestionTest {
     public static final Question Q1 = new Question("title1", "contents1").writeBy(UserTest.JAVAJIGI);
     public static final Question Q2 = new Question("title2", "contents2").writeBy(UserTest.SANJIGI);
```
static 변수를 다른 테스트에서 참조하는 부분을 제거해달라고 요청해드렸던 것 같은데 그대로 있네요.

kyucumber reviewed (approved):
피드백 잘 반영해주셨습니다.

다음 단계 진행을 위해 머지하겠습니다.
